### PR TITLE
Flux followup

### PR DESCRIPTION
**Pull request by yiyixuxu**
to-do
- [x] refactor rotary embedding
   * removed Flux-specific methods/classes to use rotary embedding: i.e. `EmbedND`, `apply_rope`, `rope`; 
   * created a `FluxPosEmbed` that uses diffusers' existing `get_1d_rotary_pos_embed` method, which is already used by hunyuan dit, lumina and stable audio
   *  changed the flux transformer inputs `img_ids` and `txt_ids`: currently, these are 3d tensors with a batch dimension; I changed them to 2d since these are just positional ids that are used to encode rotary embedding, so we do not need to add a batch dimension here, and  `get_1d_rotary_pos_embed` method does not accept batched positional ids so we remove it here to be consistent across the library - note that this is a breaking change, so I made sure to deprecate it, also add a test to make sure the previous inputs will still work
- [x] refactor attention processor (combine into one, deprecated `FluxSingleAttnProcessor2_0`)

```python
# flux unit test for rotary embedding refactor

import torch
from diffusers import FluxPipeline

model_path = "black-forest-labs/FLUX.1-dev"
pipe = FluxPipeline.from_pretrained(model_path, torch_dtype=torch.bfloat16)
pipe.enable_model_cpu_offload() #save some VRAM by offloading the model to CPU. Remove this if you have enough GPU power

prompt = "A cat holding a sign that says hello world"

image = pipe(
    prompt,
    height=1024,
    width=1024,
    guidance_scale=3.5,
    num_inference_steps=50,
    max_sequence_length=512,
    generator=torch.Generator("cpu").manual_seed(0),
).images[0]

image.save(f"yiyi_test_4_out{branch}.png")
```
main            |  this PR
:-------------------------:|:-------------------------:
yiyi_test_4_out_main | yiyi_test_4_out

### CODE REVIEW

File: src/diffusers/models/transformers/transformer_flux.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # Copyright 2024 Black Forest Labs, The HuggingFace Team. All rights reserved.
+
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_flux.py:2]:
```suggestion
```

**Comment by wangqixun** [comment]
Could you please inform when the related PR for the 'flux transformer block' and 'pipeline' can be completed and merged? We are currently working on the adaptation and training of controlnet and other related plugins on the old version. After this PR is completed, we will reorganize the code to adapt to the new coding style, and then we will submit a new PR for controlnet and ipadapter.

**Comment by yiyixuxu** [comment]
@wangqixun let's not wait for this refactor to be done for the PR! we can refactor the ip-adapter and controlnet together once the PR is in

File: src/diffusers/models/attention_processor.py [hunk 2]
```diff
@@ -4105,6 +4020,17 @@ def __init__(self):
...
 
 
+class FluxSingleAttnProcessor2_0(FluxAttnProcessor2_0):
+    r"""
+    Processor for implementing scaled dot-product attention (enabled by default if you're using PyTorch 2.0).
+    """
+
+    def __init__(self):
+        deprecation_message = "`FluxSingleAttnProcessor2_0` is deprecated and will be removed in a future version. Please use `FluxAttnProcessor2_0` instead."
+        deprecate("FluxSingleAttnProcessor2_0", "1.0.0", deprecation_message)
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:4030]: Let's maybe deprecate it earlier? Not a strong opinion, though.

File: src/diffusers/models/embeddings.py [hunk 3]
```diff
@@ -540,6 +543,30 @@ def apply_rotary_emb(
         return x_out.type_as(x)
 
 
+class FluxPosEmbed(nn.Module):
```
Comment (sayakpaul) [inline src/diffusers/models/embeddings.py:546]: Maybe a reference to the original BFL inference code?

**Review by sayakpaul — APPROVED** [review]
Superb! 

So, IIUC, the batched `txt_ids` and `img_ids` are not at a necessity now because of how we're doing the RoPE in the refactored class (`FluxPosEmbed`)? Or are there any additional differences to be aware of?

File: src/diffusers/models/embeddings.py [hunk 4]
```diff
@@ -540,6 +543,31 @@ def apply_rotary_emb(
...
+        self.theta = theta
+        self.axes_dim = axes_dim
+
+    def forward(self, ids: torch.Tensor) -> torch.Tensor:
+        n_axes = ids.shape[-1]
+        cos_out = []
+        sin_out = []
+        pos = ids.squeeze().float().cpu().numpy()
+        is_mps = ids.device.type == "mps"
+        freqs_dtype = torch.float32 if is_mps else torch.float64
```
Comment (yiyixuxu) [inline src/diffusers/models/embeddings.py:559]: @sayakpaul the results for flux are identical with this refactor 
the only other difference is here, where we downcast the dtype for mps see https://github.com/huggingface/diffusers/pull/9133 for more details

Comment (sayakpaul) [reply src/diffusers/models/embeddings.py:559]: Aye, thanks!

File: src/diffusers/pipelines/flux/pipeline_flux.py [hunk 5]
```diff
@@ -364,8 +360,7 @@ def encode_prompt(
                 unscale_lora_layers(self.text_encoder_2, lora_scale)
 
         dtype = self.text_encoder.dtype if self.text_encoder is not None else self.transformer.dtype
-        text_ids = torch.zeros(batch_size, prompt_embeds.shape[1], 3).to(device=device, dtype=dtype)
```
Comment (DN6) [inline src/diffusers/pipelines/flux/pipeline_flux.py:367]: Nice 👍🏽

**Review by DN6 — APPROVED** [review]
LGTM 👍🏽

**Pull request merged**